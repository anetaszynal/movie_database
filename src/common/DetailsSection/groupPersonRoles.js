export const groupPersonRoles = (credits, role) => role === 'character'
? credits
: Object.values(credits.reduce((c, {name,job, profile_path, credit_id, id}) => ({
    ...c,
    [name]: {
      name,
      profile_path,
      credit_id,
      id,
      job: [
        ...(
          c[name]?.job || []
        ), job
      ]
    }
  }),{},));