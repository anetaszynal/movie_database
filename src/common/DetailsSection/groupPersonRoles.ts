import { MovieCreditsCast, MovieCreditsCrew } from "../../models/credits.model";

export type CrewPersonRoles = {name: string,profile_path: string | null, credit_id: string, id: number, job: string[]}

export const groupPersonRoles = (credits: MovieCreditsCrew[] | MovieCreditsCast[], role:'character' | 'job') => role === 'character'
? credits as MovieCreditsCast[]
: Object.values((credits as MovieCreditsCrew[]).reduce((c: MovieCreditsCrew & Record<string, CrewPersonRoles>, { name, job, profile_path, credit_id, id }) => ({
    ...c,
    [name]: {
      name,
      profile_path,
      credit_id,
      id,
      job: [
        ...(
          c[name]?.job || []
        ), 
        job
      ]
    }
  }),{} as MovieCreditsCrew & Record<string, CrewPersonRoles>));