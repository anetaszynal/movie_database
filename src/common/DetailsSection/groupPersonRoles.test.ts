import { groupPersonRoles } from "./groupPersonRoles";

describe('groupPersonRoles', () => {
  const movieCreditsCrewData = [
    {
        "adult": false,
        "gender": 2,
        "id": 10980,
        "known_for_department": "Acting",
        "name": "Daniel Radcliffe",
        "original_name": "Daniel Radcliffe",
        "popularity": 29.485,
        "profile_path": "/f9WKorjfanW4PxTxhjRvHtCmfKf.jpg",
        "cast_id": 22,
        "character": "Harry Potter",
        "credit_id": "52fe4268c3a36847f801c4db",
        "order": 0
    },
    {
        "adult": false,
        "gender": 2,
        "id": 10989,
        "known_for_department": "Acting",
        "name": "Rupert Grint",
        "original_name": "Rupert Grint",
        "popularity": 12.605,
        "profile_path": "/iFlkpTaOF6fGLqxz8b0PhI0i0zN.jpg",
        "cast_id": 23,
        "character": "Ron Weasley",
        "credit_id": "52fe4268c3a36847f801c4df",
        "order": 1
    },
  ]

  const movieCreditsCastData = [
    {
        "adult": false,
        "gender": 2,
        "id": 1129,
        "known_for_department": "Camera",
        "name": "Slawomir Idziak",
        "original_name": "Slawomir Idziak",
        "popularity": 0.6,
        "profile_path": null,
        "credit_id": "52fe4268c3a36847f801c489",
        "department": "Camera",
        "job": "Director of Photography"
    },
    {
        "adult": false,
        "gender": 2,
        "id": 1129,
        "known_for_department": "Camera",
        "name": "Slawomir Idziak",
        "original_name": "Slawomir Idziak",
        "popularity": 0.6,
        "profile_path": null,
        "credit_id": "52fe4268c3a36847f801c49b",
        "department": "Camera",
        "job": "Production Design"
    },
  ]

  
  const movieCreditsCastDataExpectedResult = [
    {
        "id": 1129,
        "name": "Slawomir Idziak",
        "profile_path": null,
        "credit_id": "52fe4268c3a36847f801c49b",
        "job": ["Director of Photography", "Production Design"]
    }
  ]

  test('should group person roles together', () => {
    expect(groupPersonRoles(movieCreditsCrewData, 'character')).toMatchSnapshot(movieCreditsCrewData)
    expect(groupPersonRoles(movieCreditsCastData, 'job')).toMatchSnapshot(movieCreditsCastDataExpectedResult)
  })
})