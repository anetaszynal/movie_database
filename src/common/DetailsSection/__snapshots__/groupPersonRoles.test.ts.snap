// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupPersonRoles should group person roles together 1`] = `
Array [
  Object {
    "adult": false,
    "cast_id": 22,
    "character": "Harry Potter",
    "credit_id": "52fe4268c3a36847f801c4db",
    "gender": 2,
    "id": 10980,
    "known_for_department": "Acting",
    "name": "Daniel Radcliffe",
    "order": 0,
    "original_name": "Daniel Radcliffe",
    "popularity": 29.485,
    "profile_path": "/f9WKorjfanW4PxTxhjRvHtCmfKf.jpg",
  },
  Object {
    "adult": false,
    "cast_id": 23,
    "character": "Ron Weasley",
    "credit_id": "52fe4268c3a36847f801c4df",
    "gender": 2,
    "id": 10989,
    "known_for_department": "Acting",
    "name": "Rupert Grint",
    "order": 1,
    "original_name": "Rupert Grint",
    "popularity": 12.605,
    "profile_path": "/iFlkpTaOF6fGLqxz8b0PhI0i0zN.jpg",
  },
]
`;

exports[`groupPersonRoles should group person roles together 2`] = `
Array [
  Object {
    "credit_id": "52fe4268c3a36847f801c49b",
    "id": 1129,
    "job": Array [
      "Director of Photography",
      "Production Design",
    ],
    "name": "Slawomir Idziak",
    "profile_path": null,
  },
]
`;
